### ISSUE TITLE
Remove cloning from `impl Future for UhciFuture`

### ISSUE THREAD

**Issue by nrabulinski** [sphaerophoria/stream-os#2]
All of those approaches only work because `UhciFuture` is `Unpin`. If it wasn't, `DerefMut` wouldn't be implemented on `Pin<&mut Self>`, thus we'd need to implement projections for those fields.

```rust
// Approach 1: Iterate over indexes (only works because the ids are Copy, otherwise we'd need to clone them anyway)
let mut ret = Vec::new();
for i in 0..self.ids.len() {
    let id = self.ids[i];
    let mut buf = self.buffers.remove(&id).expect("Failed to remove id");
    buf.hw_sync();
    ret.push(buf.buf);
}

// Approach 2: Implicitly borrow everything mutably
// ids is &mut Vec<u64>
// buffers is &mut &mut BTreeMap<u64, Box<TransferDescriptorStorage>>
let UhciFuture { ids, buffers, .. } = &mut *self;
let mut ret = Vec::new();
for id in ids {
    let mut buf = buffers.remove(id).expect("Failed to remove id");
    buf.hw_sync();
    ret.push(buf.buf);
}

// Approach 3: Explicitly borrow what we need
// ids is &Vec<u64>
// buffers is &mut &mut BTreeMap<u64, Box<TransferDescriptorStorage>>
let UhciFuture {
    ref ids,
    ref mut buffers,
    ..
} = *self;
let mut ret = Vec::new();
for id in ids {
    let mut buf = buffers.remove(id).expect("Failed to remove id");
    buf.hw_sync();
    ret.push(buf.buf);
}
```

This works because when destructuring, we're creating both of the references *at the same time* which tells the borrow checker that nothing will happen between borrowing one versus borrowing the other field, whereas (currently) the borrow checker doesn't know that the borrows are disjoint if you create them separately, like by doing `let ids = &self.ids; let buffers = &mut self.buffers;`

**Comment by sphaerophoria** [comment]
Thanks, good ideas. I went for something similar to approach 2 but just didn't destructure

```
        let self_unpin = &mut *self;
        let mut ret = Vec::new();
        for id in &self_unpin.ids {
            let mut buf = self_unpin.buffers.remove(id).expect("Failed to remove id");
            buf.hw_sync();
            ret.push(buf.buf);
        }
```

**Comment by sphaerophoria** [comment]
Fixed in 751846b3a24e0988d01d7f0ac94b18c4636a7a6d
